File: src/prod/src/data/txnreplicator/loggingreplicator/BackupMetadataFileProperties.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace Data::LoggingReplicator;
using namespace Data::Utilities;

NTSTATUS BackupMetadataFileProperties::Create(
    __in KAllocator & allocator,
    __out SPtr & result) noexcept
{
    // Cal KGuid(__in BOOLEAN Zero = TRUE) to constructor the empty KGuid
    KGuid emptyKGuid(TRUE);
    result = _new(BACKUP_METADATA_FILE_PROPERTIES_TAG, allocator) BackupMetadataFileProperties(
        FABRIC_BACKUP_OPTION_INVALID,
        emptyKGuid,
        emptyKGuid,
        emptyKGuid,
        -1,
        TxnReplicator::Epoch::InvalidEpoch(),
        FABRIC_INVALID_SEQUENCE_NUMBER,
        TxnReplicator::Epoch::InvalidEpoch(),
        FABRIC_INVALID_SEQUENCE_NUMBER);
    
    if (result == nullptr)
    {
        return STATUS_INSUFFICIENT_RESOURCES;
    }

    if (!NT_SUCCESS(result->Status()))
    {
        return (SPtr(Ktl::Move(result)))->Status();
    }

    return STATUS_SUCCESS;
}

NTSTATUS BackupMetadataFileProperties::Create(
    __in FABRIC_BACKUP_OPTION backupOption,
    __in KGuid const & parentBackupId,
    __in KGuid const & backupId,
    __in KGuid const & partitionId,
    __in FABRIC_REPLICA_ID replicaId,
    __in TxnReplicator::Epoch const & startingEpoch,
    __in FABRIC_SEQUENCE_NUMBER startingLSN,
    __in TxnReplicator::Epoch const & backupEpoch,
    __in FABRIC_SEQUENCE_NUMBER backupLSN,
    __in KAllocator & allocator,
    __out SPtr & result) noexcept
{
    result = _new(BACKUP_METADATA_FILE_PROPERTIES_TAG, allocator) BackupMetadataFileProperties(
        backupOption,
        parentBackupId,
        backupId,
        partitionId,
        replicaId,
        startingEpoch,
        startingLSN,
        backupEpoch,
        backupLSN);

    if (result == nullptr)
    {
        return STATUS_INSUFFICIENT_RESOURCES;
    }

    if (!NT_SUCCESS(result->Status()))
    {
        return (SPtr(Ktl::Move(result)))->Status();
    }

    return STATUS_SUCCESS;
}

// Write the properties to the buffer
void BackupMetadataFileProperties::Write(__in BinaryWriter& writer)
{
    // Allow the base class to write first.    
    FileProperties::Write(writer);

    // 'option' - LONG32
    writer.Write(BackupOptionPropertyName);
    VarInt::Write(writer, static_cast<LONG32>(sizeof(LONG32)));
    writer.Write(static_cast<LONG32>(backupOption_) - backupOptionEnumOffset);

    // 'parentbackupid' - KGuid
    writer.Write(ParentBackupIdPropertyName);
    // Guid 16 bytes
    ULONG32 length = 16;
    VarInt::Write(writer, length);
    writer.Write(parentBackupId_);

    // 'backupid' - KGuid
    writer.Write(BackupIdPropertyName);
    VarInt::Write(writer, length);
    writer.Write(backupId_);

    // 'partitionid' - KGuid
    writer.Write(PartitionIdPropertyName);
    VarInt::Write(writer, length);
    writer.Write(partitionId_);

    // 'replicaid' - LONG64
    writer.Write(ReplicaIdPropertyName);
    VarInt::Write(writer, static_cast<ULONG32>(sizeof(LONG64)));
    writer.Write(replicaId_);

    // 'epoch' - Epoch
    writer.Write(StartingEpochPropertyName);
    VarInt::Write(writer, static_cast<ULONG32>(sizeof(LONG64) * 2));
    writer.Write(startingEpoch_.DataLossVersion);
    writer.Write(startingEpoch_.ConfigurationVersion);

    // 'lsn' - FABRIC_SEQUENCE_NUMBER
    writer.Write(StartLSNPropertyName);
    VarInt::Write(writer, static_cast<ULONG32>(sizeof(FABRIC_SEQUENCE_NUMBER)));
    writer.Write(startingLSN_);

    // 'epoch' - Epoch
    writer.Write(BackupEpochPropertyName);
    VarInt::Write(writer, static_cast<ULONG32>(sizeof(LONG64) * 2));
    writer.Write(backupEpoch_.DataLossVersion);
    writer.Write(backupEpoch_.ConfigurationVersion);

    // 'lsn' - FABRIC_SEQUENCE_NUMBER
    writer.Write(BackupLSNPropertyName);
    VarInt::Write(writer, static_cast<ULONG32>(sizeof(FABRIC_SEQUENCE_NUMBER)));
    writer.Write(backupLSN_);
}

// Read the properties from the buffer
void BackupMetadataFileProperties::ReadProperty(
    __in BinaryReader & reader, 
    __in KStringView const & propertyName,
    __in ULONG32 valueSize)
{
    // Properties are key-value pairs, so read in corresponding properties
    if (propertyName.Compare(BackupOptionPropertyName) == 0)
    {
        LONG32 backupOption;
        reader.Read(backupOption);

        backupOption_ = static_cast<FABRIC_BACKUP_OPTION>(backupOption + backupOptionEnumOffset);
    }
    else if (propertyName.Compare(ParentBackupIdPropertyName) == 0)
    {
        reader.Read(parentBackupId_);
    }
    else if (propertyName.Compare(BackupIdPropertyName) == 0)
    {
        reader.Read(backupId_);
    }
    else if (propertyName.Compare(PartitionIdPropertyName) == 0)
    {
        reader.Read(partitionId_);
    }
    else if (propertyName.Compare(ReplicaIdPropertyName) == 0)
    {
        reader.Read(replicaId_);
    }
    else if (propertyName.Compare(StartingEpochPropertyName) == 0)
    {
        LONG64 dataLossNumber = FABRIC_INVALID_SEQUENCE_NUMBER;
        LONG64 configurationNumber = FABRIC_INVALID_SEQUENCE_NUMBER;
        reader.Read(dataLossNumber);
        reader.Read(configurationNumber);
        startingEpoch_ = TxnReplicator::Epoch(dataLossNumber, configurationNumber);
    }
    else if (propertyName.Compare(StartLSNPropertyName) == 0)
    {
        reader.Read(startingLSN_);
    }
    else if (propertyName.Compare(BackupEpochPropertyName) == 0)
    {
        LONG64 dataLossNumber = FABRIC_INVALID_SEQUENCE_NUMBER;
        LONG64 configurationNumber = FABRIC_INVALID_SEQUENCE_NUMBER;
        reader.Read(dataLossNumber);
        reader.Read(configurationNumber);
        backupEpoch_ = TxnReplicator::Epoch(dataLossNumber, configurationNumber);
    }
    else if (propertyName.Compare(BackupLSNPropertyName) == 0)
    {
        reader.Read(backupLSN_);
    }
    else
    {
        // If the properties is unknown, just skip it.
        FileProperties::ReadProperty(reader, propertyName, valueSize);
    }
}

TxnReplicator::Epoch BackupMetadataFileProperties::get_BackupEpoch() const { return backupEpoch_; }
void BackupMetadataFileProperties::put_BackupEpoch(__in TxnReplicator::Epoch const & epoch) { backupEpoch_ = epoch; }

KGuid BackupMetadataFileProperties::get_BackupId() const { return backupId_; }
void BackupMetadataFileProperties::put_BackupId(__in KGuid const & backupId) { backupId_ = backupId; }

FABRIC_SEQUENCE_NUMBER BackupMetadataFileProperties::get_BackupLSN() const { return backupLSN_; }
void BackupMetadataFileProperties::put_BackupLSN(__in FABRIC_SEQUENCE_NUMBER backupLSN) { backupLSN_ = backupLSN; }

FABRIC_BACKUP_OPTION BackupMetadataFileProperties::get_BackupOption() const { return backupOption_; }
void BackupMetadataFileProperties::put_BackupOption(__in FABRIC_BACKUP_OPTION backupOption) { backupOption_ = backupOption; }

KGuid BackupMetadataFileProperties::get_ParentBackupId() const { return parentBackupId_; }
void BackupMetadataFileProperties::put_ParentBackupId(__in KGuid const & parentBackupId) { parentBackupId_ = parentBackupId; }

KGuid BackupMetadataFileProperties::get_PartitionId() const { return partitionId_; }
void BackupMetadataFileProperties::put_PartitionId(__in KGuid const & partitionId) { partitionId_ = partitionId; }

FABRIC_REPLICA_ID BackupMetadataFileProperties::get_ReplicaId() const { return replicaId_; }
void BackupMetadataFileProperties::put_ReplicaId(__in FABRIC_REPLICA_ID replicaId) { replicaId_ = replicaId; }

TxnReplicator::Epoch BackupMetadataFileProperties::get_StartingEpoch() const { return startingEpoch_; }
void BackupMetadataFileProperties::put_StartingEpoch(__in TxnReplicator::Epoch const & epoch) { startingEpoch_ = epoch; }

FABRIC_SEQUENCE_NUMBER BackupMetadataFileProperties::get_StartingLSN() const { return startingLSN_; }
void BackupMetadataFileProperties::put_StartingLSN(__in FABRIC_SEQUENCE_NUMBER startingLSN) { startingLSN_ = startingLSN; }

BackupMetadataFileProperties::BackupMetadataFileProperties(
    __in FABRIC_BACKUP_OPTION backupOption,
    __in KGuid const & parentBackupId,
    __in KGuid const & backupId,
    __in KGuid const & partitionId,
    __in FABRIC_REPLICA_ID replicaId,
    __in TxnReplicator::Epoch const & startingEpoch,
    __in FABRIC_SEQUENCE_NUMBER startingLSN,
    __in TxnReplicator::Epoch const & backupEpoch,
    __in FABRIC_SEQUENCE_NUMBER backupLSN) noexcept
    : backupOption_(backupOption)
    , parentBackupId_(parentBackupId)
    , backupId_(backupId)
    , partitionId_(partitionId)
    , replicaId_(replicaId)
    , startingEpoch_(startingEpoch)
    , startingLSN_(startingLSN)
    , backupEpoch_(backupEpoch)
    , backupLSN_(backupLSN)
{
}

BackupMetadataFileProperties::~BackupMetadataFileProperties()
{
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/BackupManager/BackupMetadataFileProperties.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Fabric;
    using System.IO;
    using Microsoft.ServiceFabric.Data;

    /// <summary>
    /// Properties are key-value pairs.
    /// </summary>
    internal class BackupMetadataFileProperties : FileProperties
    {
        private const string BackupEpochPropertyName = "epoch";

        private const string BackupIdPropertyName = "backupid";

        private const string BackupLsnPropertyName = "lsn";

        private const string BackupOptionPropertyName = "option";

        private const string ParentBackupIdPropertyName = "parentBackupId";

        private const string PartitionIdPropertyName = "partitionid";

        private const string ReplicaIdPropertyName = "replicaid";

        private const string StartingEpochPropertyName = "startingEpoch";

        private const string StartLsnPropertyName = "startingLsn";

        /// <summary>
        /// The epoch at the time the backup was taken.
        /// </summary>
        private Epoch backupEpoch;

        /// <summary>
        /// The backup id.
        /// </summary>
        private Guid backupId;

        /// <summary>
        /// The logical sequence number at the time the backup was taken.
        /// </summary>
        private long backupLsn;

        /// <summary>
        /// The backup option.
        /// </summary>
        private BackupOption backupOption;

        /// <summary>
        /// Parent backup id.
        /// </summary>
        private Guid parentBackupId;

        /// <summary>
        /// The partition id for the backup.
        /// </summary>
        private Guid partitionId;

        /// <summary>
        /// The replica id for the backup.
        /// </summary>
        private long replicaId;

        /// <summary>
        /// The epoch at the time the backup was taken.
        /// </summary>
        private Epoch startingEpoch;

        /// <summary>
        /// The logical sequence number at the time the backup was taken.
        /// </summary>
        private long startingLsn;

        public Epoch BackupEpoch
        {
            get { return this.backupEpoch; }

            set { this.backupEpoch = value; }
        }

        public Guid BackupId
        {
            get { return this.backupId; }

            set { this.backupId = value; }
        }

        public long BackupLsn
        {
            get { return this.backupLsn; }

            set { this.backupLsn = value; }
        }

        public BackupOption BackupOption
        {
            get { return this.backupOption; }

            set { this.backupOption = value; }
        }

        public Guid ParentBackupId
        {
            get { return this.parentBackupId; }

            set { this.parentBackupId = value; }
        }

        public Guid PartitionId
        {
            get { return this.partitionId; }

            set { this.partitionId = value; }
        }

        public long ReplicaId
        {
            get { return this.replicaId; }

            set { this.replicaId = value; }
        }

        public Epoch StartingEpoch
        {
            get { return this.startingEpoch; }

            set { this.startingEpoch = value; }
        }

        public long StartingLsn
        {
            get { return this.startingLsn; }

            set { this.startingLsn = value; }
        }

        /// <summary>
        /// Serialize the <see cref="BackupMetadataFileProperties"/> into the given stream.
        /// </summary>
        /// <param name="writer">Stream to write to.</param>
        public override void Write(BinaryWriter writer)
        {
            // Allow the base class to write first.
            base.Write(writer);

            // 'option' - uint32
            writer.Write(BackupOptionPropertyName);
            VarInt.Write(writer, (int) sizeof(uint));
            writer.Write((uint) this.BackupOption);

            // 'parentbackupid' - Guid
            writer.Write(ParentBackupIdPropertyName);
            var parentBackupIdBytes = this.ParentBackupId.ToByteArray();
            VarInt.Write(writer, (int) parentBackupIdBytes.Length);
            writer.Write(parentBackupIdBytes);

            // 'backupid' - Guid
            writer.Write(BackupIdPropertyName);
            var backupIdBytes = this.BackupId.ToByteArray();
            VarInt.Write(writer, (int) backupIdBytes.Length);
            writer.Write(backupIdBytes);

            // 'partitionid' - Guid
            writer.Write(PartitionIdPropertyName);
            var partitionIdBytes = this.PartitionId.ToByteArray();
            VarInt.Write(writer, (int) partitionIdBytes.Length);
            writer.Write(partitionIdBytes);

            // 'replicaid' - long
            writer.Write(ReplicaIdPropertyName);
            VarInt.Write(writer, (int) sizeof(long));
            writer.Write(this.ReplicaId);

            // 'epoch' - Epoch
            writer.Write(StartingEpochPropertyName);
            VarInt.Write(writer, (int) (sizeof(long)*2));
            writer.Write(this.StartingEpoch.DataLossNumber);
            writer.Write(this.StartingEpoch.ConfigurationNumber);

            // 'lsn' - long
            writer.Write(StartLsnPropertyName);
            VarInt.Write(writer, (int) sizeof(long));
            writer.Write(this.StartingLsn);

            // 'epoch' - Epoch
            writer.Write(BackupEpochPropertyName);
            VarInt.Write(writer, (int) (sizeof(long)*2));
            writer.Write(this.BackupEpoch.DataLossNumber);
            writer.Write(this.BackupEpoch.ConfigurationNumber);

            // 'lsn' - long
            writer.Write(BackupLsnPropertyName);
            VarInt.Write(writer, (int) sizeof(long));
            writer.Write(this.BackupLsn);
        }

        /// <summary>
        /// Read the given property value.
        /// </summary>
        /// <param name="reader">Stream to read from.</param>
        /// <param name="property">Property to read.</param>
        /// <param name="valueSize">The size in bytes of the value.</param>
        protected override void ReadProperty(BinaryReader reader, string property, int valueSize)
        {
            switch (property)
            {
                case BackupOptionPropertyName:
                    this.BackupOption = (BackupOption) reader.ReadUInt32();
                    break;

                case ParentBackupIdPropertyName:
                    this.ParentBackupId = new Guid(reader.ReadBytes(valueSize));
                    break;

                case BackupIdPropertyName:
                    this.BackupId = new Guid(reader.ReadBytes(valueSize));
                    break;

                case PartitionIdPropertyName:
                    this.PartitionId = new Guid(reader.ReadBytes(valueSize));
                    break;

                case ReplicaIdPropertyName:
                    this.ReplicaId = reader.ReadInt64();
                    break;

                case StartingEpochPropertyName:
                    var startingDataLossNumber = reader.ReadInt64();
                    var startingConfigurationNumber = reader.ReadInt64();
                    this.StartingEpoch = new Epoch(startingDataLossNumber, startingConfigurationNumber);
                    break;

                case StartLsnPropertyName:
                    this.StartingLsn = reader.ReadInt64();
                    break;

                case BackupEpochPropertyName:
                    var backupDataLossNumber = reader.ReadInt64();
                    var backupConfigurationNumber = reader.ReadInt64();
                    this.BackupEpoch = new Epoch(backupDataLossNumber, backupConfigurationNumber);
                    break;

                case BackupLsnPropertyName:
                    this.BackupLsn = reader.ReadInt64();
                    break;

                default:
                    base.ReadProperty(reader, property, valueSize);
                    break;
            }
        }
    }
}